### ISSUE TITLE
spack spec should warn about any preference not taken

### ISSUE THREAD

**Issue by mwkrentel** [LLNL/spack#28820]
### Summary

We discussed this at last week's meeting (2/2).  @tgamblin thought
it was a good idea.  So, I'm following up in the hope that Todd will
either write this or assign it to someone.

Motivation:

1. clingo has a lot of freedom in choosing versions and variants to
find an "optimal" solution.

2. `packages.yaml` are only preferences, not hard requirements, so it
is possible that clingo may go against some preference.

3. some packages have very long specs, so if clingo does violate some
preference, it's easy to miss.

Feature request:

`spack spec` (and probably also install) should print a warning if it
chooses a solution that violates any preference specifically requested
by the user (from packages.yaml, command line, etc).

Call a solution "ideal" (not necessarily unique) if it obeys every
preference specified by the user.  The solution from spack spec is
called "optimal," that is, a valid solution that minimizes badness.

The request is that spack should diff the optimal solution (it's
choice) with an ideal solution and warn about any choice that is not
ideal.  This much doesn't apply to default variants, unless
specifically requested.

I suggest this much should always generate a warning.

----------

For bonus points, add some option (--warn-default) that also warns
about any solution that uses a non-default version or variant, whether
explicitly requested or not.

----------

This is beyond the scope, but I think spack spec/install could use an
option that turns all soft preferences into hard requirements, so that
clingo must obey them or else fail.

----------

I'm not exactly sure what spack considers a hard requirement versus a
soft request.  I guess a `package.py` file is hard, but it surprised
me that `packages.yaml` is only a soft preference.  Not sure about the
command line.

The point of this request is that spack should be explicit about any
difference between what is requested and what is built.

### Rationale

These may be relevant, I didn't check that closely.
#27085, #26585

### Additional information

$ spack debug report
* **Spack:** 0.11.2-16373-93afaa85d3
* **Python:** 3.6.8
* **Platform:** linux-rhel8-zen2
* **Concretizer:** clingo

### General information

- [X] I have run `spack --version` and reported the version of Spack
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by kgerheiser** [comment]
Would this affect `spack concretize` because I would really like this. 

Basically, I'm trying to build a set of packages with `gcc@9.3.0` but there was a conflict in Boost, and instead of a warning or throwing an error it changed my compiler to `intel` for all packages and I had no idea why until I went through each package one by one.

**Comment by mwkrentel** [comment]
> Would this affect spack concretize because I would really like this.

Sure, I think that would be good.
And when I say, "spec", I mean both spec and solve, of course.

**Comment by alalazo** [comment]
@tgamblin @becker33 I don't think this would be extremely difficult to implement. We can analyze atoms when we rebuild a spec from the answer set and warn about any atom that has a weight different from 0 (the optimum for all our minimizations). I would wait though until the `--fresh` PR is merged so that we have a configuration file where to set an option to turn this behavior on or off.
